commercetools-convenience/src/main/java/io/sphere/sdk/queries/QueryAllImpl.java

```diff
@@ -22,42 +22,125 @@ private QueryAllImpl(final QueryDsl<T, C> baseQuery, final long pageSize) {
         this.pageSize = pageSize;
     }
 
+    /**
+     * Execute the {@link #baseQuery} on the {@code client} and return completion stage with all pages results
+     * aggregated to list.
+     * @param client the CTP client that the query is run on.
+     * @return {@link CompletionStage} with the {@link #baseQuery} results aggregated from all pages to list.
+     */
     public CompletionStage<List<T>> run(final SphereClient client) {
-        return queryPage(client, 0).thenCompose(result -> {
-            final List<CompletableFuture<List<T>>> futureResults = new ArrayList<>();
-            futureResults.add(completedFuture(result.getResults()));
-            futureResults.addAll(queryNextPages(client, result.getTotal()));
-            return transformListOfFuturesToFutureOfLists(futureResults);
+        return run(client, entry -> entry);
+    }
+
+    /**
+     * Given a {@code resultsMapper} {@link Function}, this method applies this callback on each item of the results
+     * from {@link this} instance's {@link #baseQuery} and returns a future containing a list containing the results
+     * of this {@code resultsMapper} on each item.
+     *
+     * @param client        the CTP client that the query is run on.
+     * @param resultsMapper the callback that gets called on each item of results.
+     * @param <S>           the type of the result mapped values of the callback on each requested {@code T} item.
+     * @return a future containing a list of results of the {@code resultsMapper} on each item.
+     */
+    @Nonnull
+    <S> CompletionStage<List<S>> run(final SphereClient client, final Function<T, S> resultsMapper) {
+        return queryPage(client, 0).thenApply(result -> {
```

Comment:
I personally would extract this lambda into a separate method and then use a method handle instead of the lambda. But it's a matter of taste and we can keep your code as it is 😉

Reply:
completely agree: db35382